File: sentry-android/src/main/java/com/joshdholtz/sentry/InternalStorage.java
package com.joshdholtz.sentry;

import android.content.Context;
import android.util.Log;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.StreamCorruptedException;
import java.util.ArrayList;

/**
 * Created by Hazer on 2/25/16.
 */
class InternalStorage {

    private final static String FILE_NAME = "unsent_requests";
    private ArrayList<SentryEventRequest> unsentRequests;

    static InternalStorage getInstance() {
        return LazyHolder.instance;
    }

    private static class LazyHolder {
        private static InternalStorage instance = new InternalStorage();
    }

    private InternalStorage() {
        this.unsentRequests = this.readObject(Sentry.getInstance().context);
    }

    /**
     * @return the unsentRequests
     */
    public ArrayList<SentryEventRequest> getUnsentRequests() {
        return unsentRequests;
    }

    public void addRequest(SentryEventRequest request) {
        synchronized (this) {
            Log.d(Sentry.TAG, "Adding request - " + request.uuid);
            if (!this.unsentRequests.contains(request)) {
                this.unsentRequests.add(request);
                this.writeObject(Sentry.getInstance().context, this.unsentRequests);
            }
        }
    }

    public void removeBuilder(SentryEventRequest request) {
        synchronized (this) {
            Log.d(Sentry.TAG, "Removing request - " + request.uuid);
            this.unsentRequests.remove(request);
            this.writeObject(Sentry.getInstance().context, this.unsentRequests);
        }
    }

    private void writeObject(Context context, ArrayList<SentryEventRequest> requests) {
        try {
            FileOutputStream fos = context.openFileOutput(FILE_NAME, Context.MODE_PRIVATE);
            ObjectOutputStream oos = new ObjectOutputStream(fos);
            oos.writeObject(requests);
            oos.close();
            fos.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private ArrayList<SentryEventRequest> readObject(Context context) {
        try {
            FileInputStream fis = context.openFileInput(FILE_NAME);
            ObjectInputStream ois = new ObjectInputStream(fis);
            ArrayList<SentryEventRequest> requests = (ArrayList<SentryEventRequest>) ois.readObject();
            return requests;
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (StreamCorruptedException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        return new ArrayList<SentryEventRequest>();
    }
}

File: sentry-android/src/main/java/io/vithor/sentry/raven/InternalStorage.kt
package io.vithor.sentry.raven

import android.content.Context
import android.util.Log
import java.io.*
import java.util.*

/**
 * Created by Vithorio Polten on 2/25/16.
 */
internal class InternalStorage private constructor() {
    /**
     * @return the unsentRequests
     */
    val unsentRequests: ArrayList<SentryEventRequest>

    //    private object LazyHolder {
    //        private val instance by lazy { InternalStorage() }
    //    }

    init {
        val context = Sentry.sharedClient.context
        try {
            val unsetRequestsFile = File(context?.filesDir, FILE_NAME)
            if (!unsetRequestsFile.exists()) {
                writeObject(context, ArrayList<SentryEventRequest>())
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        this.unsentRequests = this.readObject(context)
    }

    fun addRequest(request: SentryEventRequest) {
        synchronized (this) {
            Log.d(Sentry.TAG, "Adding request - " + request.uuid)
            if (!this.unsentRequests.contains(request)) {
                this.unsentRequests.add(request)
                this.writeObject(Sentry.sharedClient.context, this.unsentRequests)
            }
        }
    }

    fun removeBuilder(request: SentryEventRequest) {
        synchronized (this) {
            Log.d(Sentry.TAG, "Removing request - " + request.uuid)
            this.unsentRequests.remove(request)
            this.writeObject(Sentry.sharedClient.context, this.unsentRequests)
        }
    }

    private fun writeObject(context: Context?, requests: ArrayList<SentryEventRequest>) {
        try {
            val fos = context?.openFileOutput(InternalStorage.FILE_NAME, Context.MODE_PRIVATE)
            val oos = ObjectOutputStream(fos)
            oos.writeObject(requests)
            oos.close()
            fos?.close()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun readObject(context: Context?): ArrayList<SentryEventRequest> {
        try {
            val fis = context?.openFileInput(InternalStorage.FILE_NAME)
            val ois = ObjectInputStream(fis)
            val requests = ois.readObject() as? ArrayList<SentryEventRequest>
            ois.close()
            fis?.close()
            return requests ?: ArrayList()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: StreamCorruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }

        return ArrayList()
    }

    companion object {

        private val FILE_NAME = "unsent_requests"

        val instance: InternalStorage by lazy { InternalStorage() }
    }
}
